Source/KeyStorage/FileOneTimeKeysStorage.swift

```diff
@@ -0,0 +1,250 @@
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
+// IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+//
+// Lead Maintainer: Virgil Security Inc. <support@virgilsecurity.com>
+//
+
+import Foundation
+
+/// FileOneTimeKeysStorage errors
+///
+/// - keyAlreadyExists: This key already exists
+/// - keyNotFound: Key not found
+/// - keyAlreadyMarked: Key is already marked as orphaned
+@objc public enum FileOneTimeKeysStorageError: Int, Error {
+    case keyAlreadyExists = 1
+    case keyNotFound = 2
+    case keyAlreadyMarked = 3
+}
+
+/// One-time keys storage using file
+/// This class is thread-safe
+@objc(VSRFileOneTimeKeysStorage) open class FileOneTimeKeysStorage: NSObject, OneTimeKeysStorage {
+    private var oneTimeKeys: OneTimeKeys?
+    private let fileSystem: FileSystem
+
+    private struct OneTimeKeys: Codable {
+        var oneTimeKeys: [OneTimeKey]
+    }
+
+    /// Initializer
+    ///
+    /// - Parameter identity: identity of this user
+    @objc public init(identity: String) {
+        self.fileSystem = FileSystem(identity: identity)
+
+        super.init()
+    }
+
+    private let queue = DispatchQueue(label: "FileOneTimeKeysStorageQueue")
+    private var interactionCounter = 0
+
+    /// Starts interaction with storage
+    /// This method should be called before any other interaction with storage
+    /// This method can be called many times and works like a stack
+    ///
+    /// - Throws: Depends on implementation
+    public func startInteraction() throws {
+        try self.queue.sync {
+            if self.interactionCounter > 0 {
+                self.interactionCounter += 1
+                return
+            }
+
+            guard self.oneTimeKeys == nil else {
+                return
+            }
+
+            let data = try self.fileSystem.readOneTimeKeysFile()
+
+            if !data.isEmpty {
+                self.oneTimeKeys = try PropertyListDecoder().decode(OneTimeKeys.self, from: data)
+            }
+            else {
+                self.oneTimeKeys = OneTimeKeys(oneTimeKeys: [])
+            }
+
+            self.interactionCounter = 1
+        }
+    }
+
+    /// Stops interaction with storage
+    /// This method should be called after all interactions with storage
+    /// This method can be called many times and works like a stack
+    ///
+    /// - Throws:
+    ///         - Rethrows from PropertyListEncoder
+    ///         - Rethrows from FileSystem
+    public func stopInteraction() throws {
+        try self.queue.sync {
+            guard self.interactionCounter > 0 else {
+                fatalError("interactionCounter should be > 0")
+            }
+
+            self.interactionCounter -= 1
+
+            if self.interactionCounter > 0 {
+                return
+            }
+
+            guard let oneTimeKeys = self.oneTimeKeys else {
+                fatalError("oneTimeKeys should not be nil")
```

Comment:
fatalError is used here as logic error. This fatalError (and others) will be triggered in case of interactions with storage outside of .startInteraction() and .stopInteraction() scope. This misuse should be caught during tests and must never happen in real world.